### PR TITLE
New FIRST Robotics Team Info Spice

### PR DESCRIPTION
**Pull request by AlterationBrick**
https://duck.co/ia/view/first_robotics_competition_info

### CODE REVIEW

**Comment by daxtheduck** [comment]
@AlterationBrick Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please send an email to <quackslack@duckduckgo.com>.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by AlterationBrick** [comment]
This spice is at a functional stage, but I plan to work more on:
- getting the "full name" field to be more than one line so it displays all sponsors and schools
- figuring out a way to trigger on the team key, which has the structure `frcXXXX` with no space between the trigger and the team number. I dropped it for now because it would mean turning all the trigger words into regex (I think).

Thanks for giving it a look over.

File: lib/DDG/Spice/FirstRoboticsTeams.pm [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# ABSTRACT: Get info on teams in the FIRST Robotics Competition
+
+use DDG::Spice;
+
+spice is_cached => 1;
+spice proxy_cache_valid => "200 1d"; # defaults to this automatically
+
+spice wrap_jsonp_callback => 1;
+
+spice to => 'https://www.thebluealliance.com/api/v2/team/frc$1?X-TBA-App-Id=frc885:search-engine:1';
```
Comment (mattr555) [inline lib/DDG/Spice/FirstRoboticsTeams.pm]: Maybe switch out the team name for duckduckgo, just so it's more clear for TBA.

File: share/spice/first_robotics_teams/first_robotics_teams.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+(function (env) {
+    "use strict";
+    env.ddg_spice_first_robotics_teams = function(api_result) {
+
+        if (!api_result || api_result.error || api_result.length === 0 || !api_result.name) {
+            return Spice.failed('first_robotics_teams');
+        }
+        
+        // Fix the displaying of "null" for teams with no motto
+        if (api_result.motto === null) {api_result.motto = 'No motto';}
```
Comment (mattr555) [inline share/spice/first_robotics_teams/first_robotics_teams.js]: Why not remove the empty space for the motto like you did for the moreAt? This would involve building the `record_data` object outside of `Spice.add`.

File: t/FirstRoboticsTeams.t [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+use Test::More;
+use DDG::Test::Spice;
+
+spice is_cached => 1;
+
+ddg_spice_test(
+    [qw( DDG::Spice::FirstRoboticsTeams)],
+    # At a minimum, be sure to include tests for all:
+    # - primary_example_queries
+    # - secondary_example_queries
```
Comment (mattr555) [inline t/FirstRoboticsTeams.t]: You can remove these comments :)

Also you could write a `build_test` function to insert the team name and get rid of the repetitive tests. [This](https://github.com/GuiltyDolphin/zeroclickinfo-goodies/blob/port-timer/t/Timer.t#L12-L33) is a similar example.

**Comment by mattr555** [comment]
@AlterationBrick Glad you got to build this!

This looks pretty good, just some minor nitpicks noted above. Two other things:

1) A _lot_ of full names get cut off:
image
image
This is merely an observation, I can't think of a good way to trim off the sponsor names to get to the real team name (and that might not be desired anyway).

2) Is there a good way to get summative statistics for a given team. I had a perusal through the API docs and it seems like there isn't. I'm mentioning this because on the [VEX side of things](https://beta.duckduckgo.com/?q=vrc+team+765a&ia=answer) the data source used there reports this data in a more easily-consumable format. No worries if this data can't be easily retrieved from the API, just bouncing the idea because it would make the IA even more useful.

Thanks again for contributing!

**Comment by AlterationBrick** [comment]
The reason I'd like to show the full names is that they often state the actual schools the team members come from at the very end, and that might be of more interest than the sponsors.

I don't see a simple way to get summative stats. I hope to get district ranking statistics working soon, e.g. "15th in New England," but beyond that the API seems to have nothing.

Will hammer out some changes tomorrow.

**Comment by AlterationBrick** [comment]
Okay, the issues of the names getting cut off is going to sort itself out! I added one more data field for "last competed" on my local branch (it's not really working yet) and there is now a "show more" button. Clicking it expands the field with the full name.

**Comment by moollaza** [comment]
> Okay, the issues of the names getting cut off is going to sort itself out! I added one more data field for "last competed" on my local branch (it's not really working yet) and there is now a "show more" button. Clicking it expands the field with the full name.

I was about to say we **_should_** be showing a Show More in the above cases because we're trimming the table content, but it's a known bug that we're not. The team will be fixing this soon.

**Comment by mattr555** [comment]
@AlterationBrick Sorry for the delay here! Apart from the Show More thing this is almost ready:

1) The IA triggers on `top 10 frc teams`, etc. Maybe add a guard for queries like this if you plan on expanding the answer.

2) The backend returns `null` for rookie year for team 10; could we add a check for that?

image

Thanks!

@moollaza this is ready to go on beta.

**Comment by moollaza** [comment]
Deploying to Beta...

**Comment by AlterationBrick** [comment]
@mattr555 I think I can fix those two things pretty quick.

**Comment by tagawa** [comment]
Thanks @AlterationBrick. Looks like this is ready to be deployed to beta again. /cc @moollaza

**Comment by moollaza** [comment]
@AlterationBrick @tagawa updates are live on Beta!

**Comment by AlterationBrick** [comment]
Hot sauce! Thanks @tagawa, and @mattr555 for the testing.

File: lib/DDG/Spice/FirstRobotics/TeamInfo.pm [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    years_competed => {
+        is_cached => 1,
+        proxy_cache_valid => "200 1d",
+        to => 'https://www.thebluealliance.com/api/v2/team/frc$1/years_participated?X-TBA-App-Id=duckduckgo:search-engine:1'
+    }
+};
+
+triggers any => 'frc', 'frc team', 'first robotics team', 'first robotics competition';
+
+handle remainder => sub {
```
Comment (moollaza) [inline lib/DDG/Spice/FirstRobotics/TeamInfo.pm]: you'll want to use `remainder_lc` as your regex are case sensitive 👍

File: lib/DDG/Spice/FirstRobotics/TeamInfo.pm [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+# Obtain the years a team has competed
+spice alt_to => {
+    years_competed => {
+        is_cached => 1,
+        proxy_cache_valid => "200 1d",
+        to => 'https://www.thebluealliance.com/api/v2/team/frc$1/years_participated?X-TBA-App-Id=duckduckgo:search-engine:1'
+    }
+};
+
+triggers any => 'frc', 'frc team', 'first robotics team', 'first robotics competition';
```
Comment (moollaza) [inline lib/DDG/Spice/FirstRobotics/TeamInfo.pm]: let's make this `startend` -- queries with these in the middle will be much harder to validate and judging by the autocomplete `frc team ####` is a popular query format, which should cover most of the relevant queries

File: lib/DDG/Spice/FirstRobotics/TeamInfo.pm [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+handle remainder => sub {
+
+    # only continue if the query contains digits
+    return unless /\d+/;
+    
+    # eliminate irrelevant queries containing 'frc'
+    return if ($req->query_clean =~ /frc/
```
Comment (moollaza) [inline lib/DDG/Spice/FirstRobotics/TeamInfo.pm]: with `startend` we just need to verify our remainder is a number -- we can drop these checks

File: share/spice/first_robotics/team_info/first_robotics_team_info.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            return Spice.failed('first_robotics_teams');
+        }
+        
+        // Remove the space for a website if one isn't provided
+        var template_options = {
+            content: 'record',
+            moreAt: true
+        };
+
+        if (api_result.website && api_result.website != null) {
```
Comment (moollaza) [inline share/spice/first_robotics/team_info/first_robotics_team_info.js]: Though this is fine, you could also use the subtitle. You could either make the team name into the link itself, or you could do something like:

>**FIRST Robotics Competition Team 1114**
>Simbotics | Team Website

You'd need to do something like `subtitle: [ $team_name, { href: $team_website, text: "Team Website" }]`

**Comment by moollaza** [comment]
@AlterationBrick it looks like for some reason the API we're using doesn't like our API calls as I'm not seeing a result come back, however it works in DuckPAN. I'll have to have our ops team investigate

File: lib/DDG/Spice/FirstRobotics/TeamInfo.pm (hunk 5 above)
Comment (AlterationBrick) [inline lib/DDG/Spice/FirstRobotics/TeamInfo.pm]: One of my test queries is `where is frc team 811 based`, which currently triggers and serves up the result, along with other info. Just to confirm, are you advocating ignoring this type of query?

Comment (moollaza) [inline lib/DDG/Spice/FirstRobotics/TeamInfo.pm]: Hmm handling queries like that is a little touchy because we're not validating that "based" is something we know we can provide a correct answer for.

For example queries like "which is the rival team of frc 811?" or "where is frc 811 playing next" would be triggered and return a result that would be irrelevant to the query.

e.g. https://beta.duckduckgo.com/?q=rival+team+of+frc+1111&ia=reference

Actually, that could still happen with a `startend` trigger IA because we're not considering the rest of the query: 

e.g. https://beta.duckduckgo.com/?q=frc+team+885+rivals&ia=reference

Comment (moollaza) [inline lib/DDG/Spice/FirstRobotics/TeamInfo.pm]: It may be best to verify that the remainder is just a number, or a number + a keyword like: "location", "motto", "full name" etc

We can do that in a separate PR if you'd like?

**Comment by mattr555** [comment]
@moollaza Looks like it still isn't triggering on beta?

**Comment by moollaza** [comment]
@AlterationBrick sorry for the delay. This is working well on Beta! LGTM 👍
